src/expr/src/scalar/mod.rs

```diff
@@ -318,320 +318,392 @@ impl MirScalarExpr {
             MirScalarExpr::literal(e.eval(&[], temp_storage), e.typ(&relation_type).scalar_type)
         };
 
-        // Canonicalize the structure of the expression
-        self.demorgans();
-        self.undistribute_and();
-
-        self.visit_mut(&mut |e| match e {
-            MirScalarExpr::Column(_)
-            | MirScalarExpr::Literal(_, _)
-            | MirScalarExpr::CallNullary(_) => (),
-            MirScalarExpr::CallUnary { func, expr } => {
-                if expr.is_literal() {
-                    *e = eval(e);
-                } else if *func == UnaryFunc::IsNull {
-                    // (<expr1> <op> <expr2>) IS NULL can often be simplified to
-                    // (<expr1> IS NULL) OR (<expr2> IS NULL).
-                    if let MirScalarExpr::CallBinary { func, expr1, expr2 } = &mut **expr {
-                        if func.propagates_nulls() && !func.introduces_nulls() {
-                            let expr1 = expr1.take().call_unary(UnaryFunc::IsNull);
-                            let expr2 = expr2.take().call_unary(UnaryFunc::IsNull);
-                            *e = expr1.call_binary(expr2, BinaryFunc::Or);
-                        }
-                    }
-                } else if *func == UnaryFunc::Not {
-                    match &mut **expr {
-                        MirScalarExpr::CallBinary { expr1, expr2, func } => {
-                            // Transforms `NOT(a <op> b)` to `a negate(<op>) b`
-                            // if a negation exists.
-                            if let Some(negated_func) = func.negate() {
-                                *e = MirScalarExpr::CallBinary {
-                                    expr1: Box::new(expr1.take()),
-                                    expr2: Box::new(expr2.take()),
-                                    func: negated_func,
+        // 1) Simplifications that introduce and propagate constants run in a
+        //    loop until `self` no longer changes.
+        let mut old_self = MirScalarExpr::column(0);
+        while old_self != *self {
+            old_self = self.clone();
+            self.visit_mut_pre_post(
+                &mut |e| {
+                    match e {
+                        // 1a) push down IsNull expressions
```

Comment:
I think here we are pulling up subexpressions rather than pushing down. I'd simply state that we are decomposing IsNull expressions into a disjunction of simpler IsNull subexpressions